src/goto-synthesizer/cegis_verifier.cpp

```diff
@@ -0,0 +1,529 @@
+  options.set_option("symex-cache-dereferences", false);
+  options.set_option("rewrite-union", true);
+  options.set_option("self-loops-to-assumptions", true);
+
+  // Generating trace for counterexamples.
+  options.set_option("trace", true);
+
+  // Preprocess `goto_model`. Copied from `cbmc_parse_options.cpp`.
+  remove_asm(goto_model);
+  link_to_library(
+    goto_model, log.get_message_handler(), cprover_cpp_library_factory);
+  link_to_library(
+    goto_model, log.get_message_handler(), cprover_c_library_factory);
+  process_goto_program(goto_model, options, log);
+
+  add_failed_symbols(goto_model.symbol_table);
+
+  remove_skip(goto_model);
+  label_properties(goto_model);
+  return options;
+}
+
+optionalt<loop_idt> cegis_verifiert::get_cause_loop_id(
+  const goto_tracet &goto_trace,
+  const goto_programt::const_targett violation)
+{
+  optionalt<loop_idt> result;
+
+  // build the dependence graph
+  const namespacet ns(goto_model.symbol_table);
+  dependence_grapht dependence_graph(ns);
+  dependence_graph(goto_model);
+
+  // A loop is the cause loop if the violation is dependent on the write set of
+  // the loop.
+  for(const auto &step : goto_trace.steps)
+  {
+    // Being dependent on a write set is equivalent to being dependent on one
+    // of the loop havocing instruction.
+    if(is_loop_havoc(*step.pc))
+    {
+      // Checking if `to` is dependent on `from`.
+      // `from` : loop havocing
+      // `to`   : violation
+      optionalt<goto_programt::const_targett> from;
+      optionalt<goto_programt::const_targett> to;
+
+      // Get `from`---a loop havoc instruction.
+      irep_idt from_fun_name = step.function_id;
+      const goto_functionst::goto_functiont &from_function =
+        goto_model.get_goto_function(from_fun_name);
+      for(goto_programt::const_targett it =
+            from_function.body.instructions.begin();
+          it != from_function.body.instructions.end();
+          ++it)
+      {
+        if(it->location_number == step.pc->location_number)
+        {
+          from = it;
+        }
+      }
+
+      // Get `to`---the instruction where the violation happens
+      irep_idt to_fun_name = goto_trace.get_last_step().function_id;
+      const goto_functionst::goto_functiont &to_function =
+        goto_model.get_goto_function(to_fun_name);
+      for(goto_programt::const_targett it =
+            to_function.body.instructions.begin();
+          it != to_function.body.instructions.end();
+          ++it)
+      {
+        if(it->location_number == violation->location_number)
+        {
+          to = it;
+        }
+      }
+
+      // The violation is caused by the loop havoc
+      // if it is dependent on the loop havoc.
+      if(
+        from.has_value() && to.has_value() &&
+        dependence_graph.is_flow_dependent(from.value(), to.value()))
+      {
+        result =
+          loop_idt(step.function_id, step.pc->original_loop_number.value());
```

Comment:
As a follow-up to the earlier discussion: as you do one loop at a time, would it be equally ok to do the first loop? If so, you could return early and save some compute time.